cleverhans/model_abs.py

```diff
@@ -0,0 +1,73 @@
+from abc import ABCMeta
+from keras.models import Model
+
+
+class ModelAbstraction:
+    """
+    An abstract interface for a model that exposes the required functionalities
+    """
+    __metaclass__ = ABCMeta
+
+    def __init__(self, model):
+        """
+        :param model: A function that takes a symbolic input and returns the
+                      symbolic output for the model's predictions.
+        """
+        raise NotImplementedError('Model abstraction not implementated.')
+
+    def get_hidden(self, layer):
```

Comment:
I think it should be possible to run the model once and get the state of all layers. tf.slim does something similar by returning a dictionary of `endpoints` that are basically what we want.
It may not be possible to expose hidden layers using Keras models. That's one of the reasons we moved away from using Keras as the basis for CleverHans. It's OK if your Keras implementation of the Model interface is only able to support exposing the probabilities. That way Keras models could work with most attacks like FGSM, but to use feature adversaries it'd be necessary to use something like tf.slim, Sonnet, or your own model code like in the pure_tf tutorial.